### PR TITLE
gh-126165: Improve docs of function `math.isclose`

### PR DESCRIPTION
**Pull request by Xiaokang2022**
* Issue: gh-126165

----
📚 Documentation preview 📚: https://cpython-previews--126215.org.readthedocs.build/

### CODE REVIEW

**Comment by rruuaanng** [comment]
@skirpichev

File: Doc/library/math.rst
Comment (skirpichev) [inline Doc/library/math.rst]:
```suggestion
   *abs_tol* is the minimum absolute tolerance; it must be at least ``0.0``.
   When comparing to ``0.0``, the default tolerance values means that only
   ``0.0`` is close to ``0.0``.  Pass an appropriate absolute tolerance to
   compare with ``0.0``.
```

BTW, should we keep "minimum"?

Comment (Xiaokang2022) [reply Doc/library/math.rst]: I don't think it's needed.

File: Doc/library/math.rst
Comment (skirpichev) [inline Doc/library/math.rst]: No, I don't think that nonnegative value for rel_tol does make sense.  Rather we should adjust implementation.

Comment (Xiaokang2022) [reply Doc/library/math.rst]: Sorry, I didn't read your previous comment clearly.

**Review by skirpichev — DISMISSED** [review]

**Comment by skirpichev** [comment]
CC @terryjreedy

File: Doc/library/cmath.rst
Comment (picnixz) [inline Doc/library/cmath.rst]: By `at least` here, does it mean `>=` or `>`? (people could get confused with this so I would either say "at least or equal to" or "strictly larger than" depending on the intent).

File: Doc/library/math.rst
Comment (picnixz) [inline Doc/library/math.rst]: Ditto

File: Doc/library/cmath.rst
Comment (skirpichev) [reply Doc/library/cmath.rst]: What about "*abs_tol* is the absolute tolerance; it must be nonnegative."?  Ditto for rel_tol: "*rel_tol* must be nonnegative."

FYI: PEP and docs now specify rel_tol as > 0.  While code checks >= condition instead.  I'm not sure if we should stay with code or with docs.  Nothing harmful IMO in rel_tol=0, especially if abs_tol != 0.

Comment (picnixz) [reply Doc/library/cmath.rst]: > nonnegative

I usually use that formulation to avoid the overly verbose "greater/less than or equal" so I like it.

> I'm not sure if we should stay with code or with docs

I think it's better to stick with the code because it allows people to put it to 0 if they want. Otherwise they might think "hey, rel_tol = 0 is special and I'm not allowed to do it" (or it may comes from some other computed value). Saying that both `abs_tol` and `rel_tol` must be nonnegative is simpler (and correct) for learning how to use that function.

---

cc @willingc

File: Doc/library/cmath.rst
Comment (willingc) [inline Doc/library/cmath.rst]: Hmm... I'm not sure that this wording clarifies the behavior.
```suggestion
   When comparing to ``0.0``, the relative tolerance value means
```

Comment (willingc) [reply Doc/library/cmath.rst]: I would make a similar change in the other file too.

**Review by willingc — APPROVED** [review]
Thanks @Xiaokang2022 for the PR and folks for reviews.

File: Doc/library/cmath.rst
Comment (willingc) [reply Doc/library/cmath.rst]: @skirpichev Please re-review. Now that I'm reflecting on this, I'm not sure my suggestion about `relative` is correct. Perhaps it should have been `default absolute`. Thanks!

Comment (Xiaokang2022) [reply Doc/library/cmath.rst]: I think it should be `default absolute`. My opinion may be wrong and is just for reference.

Comment (skirpichev) [reply Doc/library/cmath.rst]: > the default tolerance values

The point was that defaults (both for absolute=0.0 and relative tolerances) means that only 0.0 is close to 0.0.  Maybe this: "*abs_tol* is the absolute tolerance; it must be nonnegative.  Only ``0.0` is close to ``0.0`` per default.  Pass an appropriate absolute tolerance to compare with ``0.0``."

Comment (willingc) [reply Doc/library/cmath.rst]: Thanks @skirpichev and @Xiaokang2022. Let's use @skirpichev's suggested wording. Thanks!

Comment (Xiaokang2022) [reply Doc/library/cmath.rst]: @skirpichev @willingc Please review again. Thanks!

**Review by willingc — APPROVED** [review]

File: Doc/library/cmath.rst
Comment (terryjreedy) [reply Doc/library/cmath.rst]: I am responding in a fresh comment in which I will request changes.  Please do not merge before that.

File: Doc/library/cmath.rst
Comment (terryjreedy) [reply Doc/library/cmath.rst]: I agree that the doc should be changed to acknowledge rel_tol=0.0.  As noted on both the issue and above, the effect is to make isclose depend only on abs_tol.  This could be mentioned in the rel_tol sentence.  Perhaps this is why the code deviated from the PEP.

File: Doc/library/cmath.rst
Comment (terryjreedy) [inline Doc/library/cmath.rst]: The initial PR is my suggestion in https://github.com/python/cpython/issues/126165#issuecomment-2448116678 that replaced 2 sentences with 3:
"*abs_tol* is the minimum absolute tolerance; it must be at least ``0.0``.
When comparing to ``0.0``, the default *abs_tol* of ``0.0`` means that only ``0.0`` is close to ``0.0``.
Pass a larger tolerance to allow for floating point inaccuracies."
I respond to several comments elsewhere by discussing changes to each sentence separately and finish with a suggested rewrite of all 3.

Sentence 1: removing 'minimum', carried over from the original, looks fine to me.  'nonnegative' is equivalent to 'at least 0.0'.  In English, the latter means 'equal to or greater than 0.0' ('>= 0.0').  Preference here is partly a matter of taste and judgment of what is better for readers with a non-English first language.

Sentence 2: Since `isclose(x, 0)` is computed as `abs(x) <= rel_tol * abs(x)`; the result is False for x != 0 as long as rel_tol < 1.  I implicitly assumed this since if rel_tol >= 1, The result is True for any x.  The initial replacement of "abs_tol of 0.0" by "tolerance values" acknowledges dependence on a sane rel_tol at the cost of removing the default and being more confusing.  Specifying the default is part of the purpose of the sentence as I wrote it.  I think my original phrase was better.  I also dislike the replacement sentence "Only ``0.0`` is close to ``0.0`` per default." To me, "per default" is an awkward legalism.

Sentence 3: My suggest is a bit awkward.  In the suggested replacement, "Pass an appropriate absolute tolerance to compare with ``0.0``.", 'positive' is missing and 'to compare with 0.0' is redundant as this is the context started in the sentence before.

Here is a new abs_tol entry that takes into account the comments above.  Among other things, it separates statement of the default from discussion of closeness to 0.   I will work best with changes elsewhere that I will propose in my review comment. The markup may need tweaks.
```suggestion
   *abs_tol* is the absolute tolerance; it defaults to ``0.0`` and it must be nonnegative.  When comparing ``x`` to ``0.0``, ``isclose(x, 0)`` is computed as ``abs(x) <= rel_tol  * abs(x)``, which is false for any ``x`` and rel_tol less than ``1.0``.  So add an appropriate positive abs_tol argument to the call. 
```

**Review by terryjreedy — CHANGES_REQUESTED** [review]
I would like 3 changes.

1. move the result sentence "If no errors occur, ... abs_tol)``." to immediately after the second sentence "Whether ... is determined ... [by] tolerances."  The result sentence specifies how the tolerances determine, which is needed to really understand the entries that follow.
2. At the end of the rel_tol entry, either replace "greater than zero" with "nonnegative"  or prefix it with "equal or ".  Either way, add "and less than one" at the end.  The meaning of the function assumes much less than one.
3. Further revise the abs_tol entry as suggested below in my comment.

File: Doc/library/cmath.rst
Comment (terryjreedy) [reply Doc/library/cmath.rst]: With respect to point 2, I think rel_tol >= 1 is likely a mistake of entering a percentage, such as ``1`` when ``.01`` was meant.  So we might say that it must be nonnegative and normally is much less than 1.

**Comment by Xiaokang2022** [comment]
@terryjreedy, I have made the requested changes; please review again.

File: Doc/library/cmath.rst
Comment (Xiaokang2022) [reply Doc/library/cmath.rst]: I've made a few slight formatting changes to your suggestion:

```diff
- less than one
+ less than ``1.0``
```

```diff
- which is false for
+ which is ``False`` for
```

**Review by terryjreedy — APPROVED** [review]

File: Doc/library/cmath.rst
Comment (terryjreedy) [reply Doc/library/cmath.rst]: A change was made.

File: Doc/library/cmath.rst
Comment (terryjreedy) [reply Doc/library/cmath.rst]: Looks good.  I will wait a day before merging.

**Review by willingc — APPROVED** [review]
Great improvements over the original. Thanks @terryjreedy for the additional content.

**Pull request merged**
